src/main/java/reposense/git/GitClone.java

```diff
@@ -59,4 +63,53 @@ private static void clone(RepoLocation location, String repoName) throws IOExcep
         Files.createDirectories(rootPath);
         runCommand(rootPath, "git clone " + addQuote(location.toString()));
     }
+
+    /**
+     * Spawns a process to clone repo specified in {@code repoConfig}. Does not wait for process to finish executing.
+     * Can only handle a maximum of one spawned process at any time and subsequent calls are ignored if current process
+     * is still running.
+     */
+    public static void spawnCloneProcess(RepoConfiguration repoConfig) throws GitCloneException {
+        if (crp != null) {
+            return;
+        }
+        try {
+            FileUtil.deleteDirectory(repoConfig.getRepoRoot());
+            logger.info("Cloning in parallel from " + repoConfig.getLocation() + "...");
+            Path rootPath = Paths.get(FileUtil.REPOS_ADDRESS, repoConfig.getRepoFolderName());
+            Files.createDirectories(rootPath);
+            crp = runCommandAsync(rootPath, "git clone " + addQuote(repoConfig.getLocation().toString()));
+        } catch (RuntimeException rte) {
+            logger.log(Level.SEVERE, "Error encountered in Git Cloning, will attempt to continue analyzing", rte);
+            throw new GitCloneException(rte);
+        } catch (IOException ioe) {
+            throw new GitCloneException(ioe);
+        }
+    }
+
+    /**
+     * Waits for previously spawned clone process to finish executing.
+     * Should only be called after {@code spawnCloneProcess} has been called.
+     */
+    public static void waitForCloneProcess(RepoConfiguration repoConfig) throws GitCloneException {
+        try {
+            crp.waitForProcess();
+            logger.info("Cloning of " + repoConfig.getLocation() + " completed!");
+        } catch (RuntimeException | CommandRunnerProcessException e) {
+            crp = null;
+            logger.log(Level.SEVERE, "Error encountered in Git Cloning, will attempt to continue analyzing", e);
+            throw new GitCloneException(e);
+        }
+        crp = null;
+
+        try {
+            if (repoConfig.getBranch().equals(RepoConfiguration.DEFAULT_BRANCH)) {
```

Comment:
perhaps extract a method for these lines and place it as a member method in RepoConfiguration.

```
        try {
            if (repoConfig.getBranch().equals(RepoConfiguration.DEFAULT_BRANCH)) {
                String currentBranch = GitBranch.getCurrentBranch(repoConfig.getRepoRoot());
                repoConfig.setBranch(currentBranch);
            }
        } catch (RuntimeException e) {
            logger.log(Level.SEVERE, "Branch does not exist! Analysis terminated.", e);
            throw new GitCloneException(e);
        }
```

Can call it `renewBranch` or `fetchBranch` etc.

Reply:
Ok makes sense. It might be weird for `RepoConfiguration` to throw `GitCloneException`, so I'll handle the exception at a higher level.

Reply:
You can create its own exception, e.g. `BranchNotFoundException`. The respective function that calls it can catch it and decide what to do afterward. e.g. catch `BranchNotFoundException` and throw `
GitCloneException` etc